numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -468,6 +471,18 @@ PyArray_ConcatenateArrays(int narrays, PyArrayObject **arrays, int axis,
             return NULL;
         }
 
+        initial_byteorder = PyArray_DESCR(arrays[0])->byteorder;
+        for (iarrays = 0; iarrays < narrays; ++iarrays) {
+            iterative_byteorder = PyArray_DESCR(arrays[iarrays])->byteorder;
+            if (iterative_byteorder != initial_byteorder) {
+                single_byteorder = 0;
+            }
+        }
+
+        if (single_byteorder == 1 && dtype == NULL) {
+            descr->byteorder = initial_byteorder;
+        }
```

Comment:
These 3 lines are the ones that cause somewhat-stochastic test failures in the full suite, that in my hands are not reproducible when testing those individuals tests (or even their entire modules) in isolation on this branch. If they are commented out, then the new test I added fails, but everything else passes.

Reference counting or something else?

Reply:
The problem is that `descr` is a singleton (or can be).  You would have to create a new dtype and swap bytes.  (There should be a `NewByteOrder` function for that I think).